src/main.cpp

```diff
@@ -123,12 +125,51 @@ void hideConsoleWindow() {
 }
 
 void start_viewer(string map) {
-	if (!fileExists(map)) {
+	if (map != "-testnew" && !fileExists(map)) {
 		logf("ERROR: File not found: %s", map.c_str());
 		return;
 	}
 	Renderer renderer = Renderer();
-	renderer.addMap(new Bsp(map));
+	if (map == "-testnew")
+	{
+		logf("Create new bsp: %s", map.c_str());
+		Bsp* newBsp = new Bsp();
+		newBsp->path = "testnew.bsp";
+		newBsp->name = "testnew.bsp";
+		// Create wordlspawn entity with default wad
+		Entity * tmpEntity = new Entity("worldspawn");
+		tmpEntity->addKeyvalue("wad", "halflife.wad;");
+		newBsp->ents.push_back(tmpEntity);
+		// Add and clean two textures
+		byte* tex_dat = NULL;
+		uint w, h;
+		lodepng_decode24(&tex_dat, &w, &h, aaatrigger_dat, sizeof(aaatrigger_dat));
+		newBsp->add_texture("aaatrigger", tex_dat, w, h);
+		int aaatriggerIdx = newBsp->add_texture("XCRATE7C", tex_dat, w, h);
+		newBsp->delete_embedded_textures();
+		// Make out of map is empty
+		newBsp->create_leaf(CONTENTS_EMPTY);
+		// Create test cube with 64 size
+		int size = 64;
+		vec3 mins = vec3(-size, -size, -size);
+		vec3 maxs = vec3(size, size, size);
+		newBsp->create_solid(mins,maxs, aaatriggerIdx);
+		// Set cube faces lightmap
+		BSPFACE* thisFaces = (BSPFACE*)newBsp->lumps[LUMP_FACES];
+		int thisFaceCount = newBsp->header.lump[LUMP_FACES].nLength / sizeof(BSPFACE);
+		for (int i = 0; i < thisFaceCount; i++)
+		{
+			thisFaces[i].nStyles[0] = 0;
+			thisFaces[i].nLightmapOffset = 768 * (i + 1);
+		}
+		// Update all lumps
+		newBsp->update_ent_lump();
+		newBsp->update_lump_pointers();
+		// Render map
+		renderer.addMap(newBsp);
+	}
+	else 
+		renderer.addMap(new Bsp(map));

```

Comment:
can you leave this out of the PR, or make a separate PR for this? I want to focus on fixing what's here first.

Reply:
Yes, partially.